### PR TITLE
feat(coderd): support ephemeral parameters

### PR DESCRIPTION
**Pull request by mtojek**
Related: https://github.com/coder/coder/issues/6828

This PR enables support for `ephemeral` parameters on the backend level. The value of an ephemeral parameter is not persisted between workspace builds, and it returns to default. Ephemeral parameters can't be immutable by definition (restricted by the schema in [terraform-provider-coder](https://github.com/coder/terraform-provider-coder/pull/138)). 

```tf
data "coder_parameter" "force-rebuild" {
  name         = "force-rebuild"
  type           = "bool"
  display_name = "Force rebuild"
  description  = "Rebuild the project code"
  default        = "false"
  mutable      = true
  ephemeral   = true
}
```

Follow-up issues/PRs:
* CLI support for `ephemeral` parameters
* Adjust site to support `ephemeral` parameters

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -358,10 +358,4 @@ require (
 	inet.af/peercred v0.0.0-20210906144145-0893ea02156a // indirect
 )
 
-require (
-	github.com/bep/godartsass/v2 v2.0.0 // indirect
-	github.com/hashicorp/go-plugin v1.4.4 // indirect
-	github.com/hashicorp/terraform-registry-address v0.0.0-20220623143253-7d51757b572c // indirect
-	github.com/hashicorp/terraform-svchost v0.0.0-20200729002733-f050f53b9734 // indirect
-	github.com/oklog/run v1.0.0 // indirect
```
Comment (johnstcn) [inline go.mod:366]: 🤔 Is this related to your change?

**Review by johnstcn — COMMENTED** [review]
The change looks good to me, and it appears to open up some very interesting possibilities.

I just have a few nits about naming, but I'm wary about the changes in `go.mod` and `go.sum`. As far as I can tell, this shouldn't require any dependency changes?

File: go.mod (hunk 1 above)
Comment (mtojek) [reply go.mod:366]: I'm concerned too, but if you copy `go.mod` and `go.sum` from main here, and then run `go mod tidy`, this is the result.

Comment (mafredri) [reply go.mod:366]: Perhaps we should add `tidy` as a lint/gen check?

File: codersdk/richparameters_test.go [hunk 2]
```diff
@@ -304,3 +304,89 @@ func TestRichParameterValidation(t *testing.T) {
...
+		Mutable:      true,
+		DefaultValue: "",
+		Ephemeral:    true,
+	}
+	v, err := uut.ValidateResolve(p, nil)
+	require.NoError(t, err)
+	require.Equal(t, "", v)
+}
+
+func TestParameterResolver_ValidateResolve_Ephemeral_RequiredButMissing(t *testing.T) {
```
Comment (mafredri) [inline codersdk/richparameters_test.go:376]: What about a case where required = true, but there's also a default value set? Is this possible and should we test it?

**Review by mafredri — APPROVED** [review]

File: codersdk/richparameters_test.go (hunk 2 above)
Comment (mtojek) [reply codersdk/richparameters_test.go:376]: It is not possible as `required` is a computed property, the opposite of [optional](https://github.com/coder/terraform-provider-coder/blob/main/provider/parameter.go#L105C5-L111), converted [here](https://github.com/coder/coder/blob/main/provisioner/terraform/resources.go#L500).

File: go.mod (hunk 1 above)
Comment (mtojek) [reply go.mod]: Sounds about right, but rather in a separate PR.

**Review by johnstcn — APPROVED** [review]

File: coderd/workspaces_test.go [hunk 3]
```diff
@@ -2518,6 +2518,134 @@ func TestWorkspaceWithOptionalRichParameters(t *testing.T) {
...
+		firstParameterValue = "i_am_first_parameter"
+
+		secondParameterName         = "second_parameter"
+		secondParameterType         = "string"
+		secondParameterDescription  = "This is second parameter"
+		secondParameterDefaultValue = ""
+		secondParameterMutable      = true
+		secondParameterEphemeral    = true
+
+		secondParameterValue = "i_am_ephemeral"
```
[the remark this answers is not in the document]
Comment (mtojek) [reply coderd/workspaces_test.go]: Fixed
